File: hdf-diff/test/change-event.test.ts
import { describe, it, expect } from 'vitest';
import { validateRequirementChangeEvent } from '@mitre/hdf-validators';
import { changeEventFromPrevious, type KeyState, type EventInputs } from '../src/change-event.js';

// Pinned vectors shared with the Go suite (effective_checksum_test.go).
const VECTOR_FAILED_HALF = '704f62b2d0803438ad6b7b9bab45e2c4f350b7344135a2a7f8ef986d98669021';
const VECTOR_PASSED_HALF = '73908440a3b44d76de559753babfea36987a618b80ee9d26adcf29cb5c7a5217';

function inputs(): EventInputs {
  return {
    eventId: '0190f6f2-1c4e-7c3a-9f2a-3b1d5e7a9c01',
    source: 'inspec://web01/rhel9-stig',
    sequence: 412,
    systemRef: 'apptier.hdf-system.json',
    componentId: '6e0f2a3b-9c01-4d5e-8f7a-1b2c3d4e5f60',
    requirementId: 'SV-100001',
    timestamp: '2026-07-22T14:03:11Z',
    referenceTimestamp: '2026-07-22T14:03:11Z',
  };
}

function prevFailing(): KeyState {
  return {
    effectiveStatus: 'failed',
    effectiveImpact: 0.5,
    checksum: { algorithm: 'sha256', value: VECTOR_FAILED_HALF },
  };
}

function makeReq(status: string, extra: Record<string, unknown> = {}): Record<string, unknown> {
  return {
    id: 'SV-100001',
    impact: 0.5,
    tags: {},
    descriptions: [{ label: 'default', data: 'test description' }],
    results: [{ status, codeDesc: 'test', startTime: '2025-01-01T00:00:00Z' }],
    ...extra,
  };
}

function makeOverride(extra: Record<string, unknown> = {}): Record<string, unknown> {
  return {
    type: 'waiver',
    status: 'passed',
    reason: 'approved by team lead',
    appliedBy: { type: 'simple', identifier: 'admin' },
    appliedAt: '2026-07-01T00:00:00Z',
    expiresAt: '2099-12-31T00:00:00Z',
    ...extra,
  };
}

function assertValidEvent(ev: Record<string, unknown>): void {
  const result = validateRequirementChangeEvent(ev);
  expect(result.valid, JSON.stringify(result.errors)).toBe(true);
}

describe('changeEventFromPrevious', () => {
  it('returns null when the checksum matches (no change)', async () => {
    const ev = await changeEventFromPrevious(prevFailing(), makeReq('failed'), null, inputs());
    expect(ev).toBeNull();
  });

  it('returns null when both prev and new are absent', async () => {
    expect(await changeEventFromPrevious(null, null, null, inputs())).toBeNull();
  });

  it('emits new for an unseen key (chain start)', async () => {
    const ev = await changeEventFromPrevious(null, makeReq('passed'), null, inputs());
    expect(ev).not.toBeNull();
    expect(ev!.state).toBe('new');
    expect(ev!.before).toBeNull();
    expect(ev!.priorChecksum).toBeNull();
    expect(ev!.after).not.toBeNull();
    assertValidEvent(ev!);
  });

  it('emits absent when the requirement leaves scope', async () => {
    const ev = await changeEventFromPrevious(prevFailing(), null, null, inputs());
    expect(ev).not.toBeNull();
    expect(ev!.state).toBe('absent');
    expect(ev!.after).toBeNull();
    expect(ev!.before).toEqual({ effectiveStatus: 'failed', effectiveImpact: 0.5 });
    expect((ev!.priorChecksum as Record<string, unknown>).value).toBe(VECTOR_FAILED_HALF);
    assertValidEvent(ev!);
  });

  it('classifies failed→passed as fixed with the chain link', async () => {
    const ev = await changeEventFromPrevious(prevFailing(), makeReq('passed'), null, inputs());
    expect(ev).not.toBeNull();
    expect(ev!.state).toBe('fixed');
    expect((ev!.priorChecksum as Record<string, unknown>).value).toBe(VECTOR_FAILED_HALF);
    assertValidEvent(ev!);
  });

  it('classifies passed→failed as regressed', async () => {
    const prev: KeyState = {
      effectiveStatus: 'passed',
      effectiveImpact: 0.5,
      checksum: { algorithm: 'sha256', value: VECTOR_PASSED_HALF },
    };
    const ev = await changeEventFromPrevious(prev, makeReq('failed'), null, inputs());
    expect(ev!.state).toBe('regressed');
    assertValidEvent(ev!);
  });

  it('classifies an impact-only change as updated with impactChanged', async () => {
    const req = makeReq('failed', {
      statusOverrides: [
        makeOverride({ type: 'riskAdjustment', status: undefined, impact: { value: 0.2 } }),
      ],
    });
    delete (req.statusOverrides as Record<string, unknown>[])[0].status;
    const ev = await changeEventFromPrevious(prevFailing(), req, null, inputs());
    expect(ev!.state).toBe('updated');
    expect(ev!.changeReasons).toContain('impactChanged');
    assertValidEvent(ev!);
  });

  it('echoes all envelope inputs', async () => {
    const i = { ...inputs(), schemaRef: 'https://mitre.github.io/hdf-libs/schemas/hdf-requirement-change-event/v3.4.0' };
    const ev = await changeEventFromPrevious(prevFailing(), makeReq('passed'), null, i);
    expect(ev!.eventId).toBe(i.eventId);
    expect(ev!.source).toBe(i.source);
    expect(ev!.sequence).toBe(i.sequence);
    expect(ev!.systemRef).toBe(i.systemRef);
    expect(ev!.componentId).toBe(i.componentId);
    expect(ev!.requirementId).toBe(i.requirementId);
    expect(ev!.timestamp).toBe(i.timestamp);
    expect(ev!.schemaRef).toBe(i.schemaRef);
    assertValidEvent(ev!);
  });

  it('attributes an override-driven flip when the full prior requirement is given', async () => {
    const prevReq = makeReq('failed');
    const newReq = makeReq('failed', { statusOverrides: [makeOverride()] });
    const ev = await changeEventFromPrevious(prevFailing(), newReq, prevReq, inputs());
    expect(ev!.state).toBe('fixed');
    expect(ev!.changeReasons).toContain('overrideAdded');
    expect(ev!.changeReasons).not.toContain('resultChanged');
    assertValidEvent(ev!);
  });

  it('filters batch-only reasons out of the wire vocabulary', async () => {
    const prevReq = makeReq('failed');
    const newReq = makeReq('passed', { title: 'Renamed control title' });
    const ev = await changeEventFromPrevious(prevFailing(), newReq, prevReq, inputs());
    expect(ev!.changeReasons).toContain('resultChanged');
    expect(ev!.changeReasons).not.toContain('metadataChanged');
    assertValidEvent(ev!);
  });

  it('matches the Go kernel field-for-field on the shared fixed-transition case', async () => {
    const ev = await changeEventFromPrevious(prevFailing(), makeReq('passed'), null, inputs());
    // Mirror of TestChangeEventFromPrevious_FailedToPassedIsFixed — the Go
    // suite asserts the same state/priorChecksum/envelope values.
    expect({
      state: ev!.state,
      priorChecksum: ev!.priorChecksum,
      before: ev!.before,
      sequence: ev!.sequence,
    }).toEqual({
      state: 'fixed',
      priorChecksum: { algorithm: 'sha256', value: VECTOR_FAILED_HALF },
      before: { effectiveStatus: 'failed', effectiveImpact: 0.5 },
      sequence: 412,
    });
  });
});

describe('mapped and window-dependent change reasons', () => {
  const WAIVED_CHECKSUM = 'd11c074ab9131807816013a71d986f1ceb2e5871a8a01dee4043391b7a6bf37b';

  it('maps effectiveImpactChanged to a single impactChanged', async () => {
    const prevReq = makeReq('failed');
    const newReq = makeReq('failed', {
      statusOverrides: [
        makeOverride({ type: 'riskAdjustment', impact: { value: 0.2 } }),
      ],
      effectiveImpact: 0.2,
    });
    delete (newReq.statusOverrides as Record<string, unknown>[])[0].status;

    const ev = await changeEventFromPrevious(prevFailing(), newReq, prevReq, inputs());
    expect(ev).not.toBeNull();
    expect(ev!.changeReasons).toContain('overrideAdded');
    const impactCount = (ev!.changeReasons as string[]).filter((r) => r === 'impactChanged').length;
    expect(impactCount, 'effectiveImpactChanged must map to a single impactChanged').toBe(1);
    assertValidEvent(ev!);
  });

  it('maps a base impact change to impactChanged', async () => {
    const prevReq = makeReq('failed');
    const newReq = makeReq('failed', { impact: 0.3 });
    const ev = await changeEventFromPrevious(prevFailing(), newReq, prevReq, inputs());
    expect(ev!.changeReasons).toContain('impactChanged');
    assertValidEvent(ev!);
  });

  it('detects overrideExpired across the observation window', async () => {
    // Identical scans; the waiver lapsed between observations (deliberately-past
    // expiresAt asserts expiry behavior).
    const waived = () =>
      makeReq('failed', {
        statusOverrides: [makeOverride({ expiresAt: '2026-07-21T00:00:00Z' })],
      });
    const prev: KeyState = {
      effectiveStatus: 'passed',
      effectiveImpact: 0.5,
      checksum: { algorithm: 'sha256', value: WAIVED_CHECKSUM },
    };
    const ev = await changeEventFromPrevious(prev, waived(), waived(), {
      ...inputs(),
      prevReferenceTimestamp: '2026-07-20T00:00:00Z',
      referenceTimestamp: '2026-07-22T14:03:11Z',
    });
    expect(ev).not.toBeNull();
    expect(ev!.state).toBe('regressed');
    expect(ev!.changeReasons).toContain('overrideExpired');
    expect(ev!.changeReasons).not.toContain('resultChanged');
    assertValidEvent(ev!);
  });

  it('detects overrideRemoved', async () => {
    const prevReq = makeReq('failed', { statusOverrides: [makeOverride()] });
    const newReq = makeReq('failed');
    const prev: KeyState = {
      effectiveStatus: 'passed',
      effectiveImpact: 0.5,
      checksum: { algorithm: 'sha256', value: WAIVED_CHECKSUM },
    };
    const ev = await changeEventFromPrevious(prev, newReq, prevReq, inputs());
    expect(ev!.state).toBe('regressed');
    expect(ev!.changeReasons).toContain('overrideRemoved');
    assertValidEvent(ev!);
  });
});

File: hdf-diff/go/change_event_test.go
package diff

import (
	"encoding/json"
	"testing"
	"time"

	hdf "github.com/mitre/hdf-libs/hdf-schema/dist/go/v3"
	validators "github.com/mitre/hdf-libs/hdf-validators/go/v3"
	"github.com/stretchr/testify/assert"
	"github.com/stretchr/testify/require"
)

func ceInputs() EventInputs {
	return EventInputs{
		EventID:            "0190f6f2-1c4e-7c3a-9f2a-3b1d5e7a9c01",
		Source:             "inspec://web01/rhel9-stig",
		Sequence:           412,
		SystemRef:          "apptier.hdf-system.json",
		ComponentID:        "6e0f2a3b-9c01-4d5e-8f7a-1b2c3d4e5f60",
		RequirementID:      "SV-100001",
		Timestamp:          time.Date(2026, 7, 22, 14, 3, 11, 0, time.UTC),
		ReferenceTimestamp: "2026-07-22T14:03:11Z",
	}
}

func cePrevFailing() *KeyState {
	return &KeyState{
		EffectiveStatus: "failed",
		EffectiveImpact: 0.5,
		Checksum:        hdf.Checksum{Algorithm: hdf.Sha256, Value: ecVectorFailedHalf},
	}
}

func cePassingReq() hdf.EvaluatedRequirement {
	return ceValid(stMakeRequirement(func(r *hdf.EvaluatedRequirement) {
		r.Impact = 0.5
		r.Results = []hdf.RequirementResult{stMakeResult(hdf.Passed)}
	}))
}

// ceValid fills the schema-required minimums the shared builders omit
// (descriptions minItems + default label) so emitted events schema-validate.
func ceValid(req hdf.EvaluatedRequirement) hdf.EvaluatedRequirement {
	if len(req.Descriptions) == 0 {
		req.Descriptions = []hdf.Description{{Label: "default", Data: "test description"}}
	}
	return req
}

// assertValidEvent marshals the event and runs it through the real schema validator.
func assertValidEvent(t *testing.T, ev *hdf.HDFRequirementChangeEvent) {
	t.Helper()
	raw, err := json.Marshal(ev)
	require.NoError(t, err)
	result := validators.ValidateRequirementChangeEvent(raw)
	assert.True(t, result.Valid, "emitted event must be schema-valid: %v", result.Errors)
}

func TestChangeEventFromPrevious_NoChangeReturnsNil(t *testing.T) {
	req := ceValid(ecFailingReq()) // failed @ 0.5 → checksum matches cePrevFailing
	ev := ChangeEventFromPrevious(cePrevFailing(), &req, nil, ceInputs())
	assert.Nil(t, ev, "matching checksum must emit no event")
}

func TestChangeEventFromPrevious_NewKey(t *testing.T) {
	req := cePassingReq()
	ev := ChangeEventFromPrevious(nil, &req, nil, ceInputs())
	require.NotNil(t, ev)
	assert.Equal(t, hdf.EventRequirementStateNew, ev.State)
	assert.Nil(t, ev.Before, "no prior posture for a new key")
	assert.Nil(t, ev.PriorChecksum, "chain start")
	require.NotNil(t, ev.After)
	assertValidEvent(t, ev)
}

func TestChangeEventFromPrevious_AbsentKey(t *testing.T) {
	ev := ChangeEventFromPrevious(cePrevFailing(), nil, nil, ceInputs())
	require.NotNil(t, ev)
	assert.Equal(t, hdf.EventRequirementStateAbsent, ev.State)
	assert.Nil(t, ev.After, "absent has no after by definition")
	before, ok := ev.Before.(map[string]interface{})
	require.True(t, ok, "before must be the thin projection")
	assert.Equal(t, "failed", before["effectiveStatus"])
	assert.InDelta(t, 0.5, before["effectiveImpact"].(float64), 1e-9)
	prior, ok := ev.PriorChecksum.(map[string]interface{})
	require.True(t, ok)
	assert.Equal(t, ecVectorFailedHalf, prior["value"])
	assertValidEvent(t, ev)
}

func TestChangeEventFromPrevious_BothNilReturnsNil(t *testing.T) {
	assert.Nil(t, ChangeEventFromPrevious(nil, nil, nil, ceInputs()))
}

func TestChangeEventFromPrevious_FailedToPassedIsFixed(t *testing.T) {
	req := cePassingReq()
	ev := ChangeEventFromPrevious(cePrevFailing(), &req, nil, ceInputs())
	require.NotNil(t, ev)
	assert.Equal(t, hdf.EventRequirementStateFixed, ev.State)
	prior, ok := ev.PriorChecksum.(map[string]interface{})
	require.True(t, ok)
	assert.Equal(t, ecVectorFailedHalf, prior["value"], "priorChecksum links the chain")
	before, ok := ev.Before.(map[string]interface{})
	require.True(t, ok)
	assert.Equal(t, map[string]interface{}{"effectiveStatus": "failed", "effectiveImpact": 0.5}, before)
	assertValidEvent(t, ev)
}

func TestChangeEventFromPrevious_PassedToFailedIsRegressed(t *testing.T) {
	prev := &KeyState{
		EffectiveStatus: "passed",
		EffectiveImpact: 0.5,
		Checksum:        hdf.Checksum{Algorithm: hdf.Sha256, Value: ecVectorPassedHalf},
	}
	req := ceValid(ecFailingReq())
	ev := ChangeEventFromPrevious(prev, &req, nil, ceInputs())
	require.NotNil(t, ev)
	assert.Equal(t, hdf.EventRequirementStateRegressed, ev.State)
	assertValidEvent(t, ev)
}

func TestChangeEventFromPrevious_ImpactOnlyChangeIsUpdated(t *testing.T) {
	// Status stays failed; a risk adjustment drops effective impact.
	req := ceValid(ecFailingReq())
	req.StatusOverrides = []hdf.StatusOverride{{
		Type:      hdf.RiskAdjustment,
		Impact:    &hdf.ImpactOverride{Value: 0.2},
		Reason:    "environmental adjustment",
		AppliedBy: hdf.Identity{Type: hdf.Simple, Identifier: "admin"},
		AppliedAt: time.Date(2026, 7, 1, 0, 0, 0, 0, time.UTC),
		ExpiresAt: time.Date(2099, 12, 31, 0, 0, 0, 0, time.UTC),
	}}
	ev := ChangeEventFromPrevious(cePrevFailing(), &req, nil, ceInputs())
	require.NotNil(t, ev)
	assert.Equal(t, hdf.EventRequirementStateUpdated, ev.State, "same status, changed impact → updated")
	assert.Contains(t, ev.ChangeReasons, hdf.EventChangeReasonImpactChanged,
		"impact delta is provable from thin prev state")
	assertValidEvent(t, ev)
}

func TestChangeEventFromPrevious_EnvelopeEchoesInputs(t *testing.T) {
	req := cePassingReq()
	in := ceInputs()
	schemaRef := "https://mitre.github.io/hdf-libs/schemas/hdf-requirement-change-event/v3.4.0"
	in.SchemaRef = schemaRef
	ev := ChangeEventFromPrevious(cePrevFailing(), &req, nil, in)
	require.NotNil(t, ev)
	assert.Equal(t, in.EventID, ev.EventID)
	assert.Equal(t, in.Source, ev.Source)
	assert.Equal(t, in.Sequence, ev.Sequence)
	assert.Equal(t, in.SystemRef, ev.SystemRef)
	assert.Equal(t, in.ComponentID, ev.ComponentID)
	assert.Equal(t, in.RequirementID, ev.RequirementID)
	assert.Equal(t, in.Timestamp, ev.Timestamp)
	require.NotNil(t, ev.SchemaRef)
	assert.Equal(t, schemaRef, *ev.SchemaRef)
	assertValidEvent(t, ev)
}

func TestChangeEventFromPrevious_RicherReasonsWithPrevRequirement(t *testing.T) {
	// Waiver added: failed → passed via override. With the full prior requirement
	// available, the classifier attributes the flip to the override, not a rescan.
	prevReq := ceValid(ecFailingReq())
	newReq := ceValid(ecFailingReq())
	newReq.StatusOverrides = []hdf.StatusOverride{stMakeOverride(struct {
		Type      string
		Status    hdf.ResultStatus
		Reason    string
		AppliedAt time.Time
		ExpiresAt time.Time
	}{Type: "waiver", Status: hdf.Passed})}

	ev := ChangeEventFromPrevious(cePrevFailing(), &newReq, &prevReq, ceInputs())
	require.NotNil(t, ev)
	assert.Equal(t, hdf.EventRequirementStateFixed, ev.State)
	assert.Contains(t, ev.ChangeReasons, hdf.EventChangeReasonOverrideAdded)
	assert.NotContains(t, ev.ChangeReasons, hdf.EventChangeReasonResultChanged,
		"results are identical; the override is the cause")
	assertValidEvent(t, ev)
}

func TestChangeEventFromPrevious_FiltersNonEventReasons(t *testing.T) {
	// A title change makes the batch classifier emit metadataChanged, which is
	// NOT in the event vocabulary and must be filtered out.
	prevReq := ceValid(ecFailingReq())
	newReq := cePassingReq()
	newReq.Title = stPtrString("Renamed control title")

	ev := ChangeEventFromPrevious(cePrevFailing(), &newReq, &prevReq, ceInputs())
	require.NotNil(t, ev)
	for _, r := range ev.ChangeReasons {
		assert.NotEqual(t, "metadataChanged", string(r),
			"batch-only reasons must not reach the wire")
	}
	assert.Contains(t, ev.ChangeReasons, hdf.EventChangeReasonResultChanged)
	assertValidEvent(t, ev)
}

func TestChangeEventFromPrevious_MappedImpactReasons(t *testing.T) {
	t.Run("effectiveImpactChanged maps to impactChanged exactly once", func(t *testing.T) {
		prevReq := ceValid(ecFailingReq())
		newReq := ceValid(ecFailingReq())
		newReq.StatusOverrides = []hdf.StatusOverride{{
			Type:      hdf.RiskAdjustment,
			Impact:    &hdf.ImpactOverride{Value: 0.2},
			Reason:    "environmental adjustment",
			AppliedBy: hdf.Identity{Type: hdf.Simple, Identifier: "admin"},
			AppliedAt: time.Date(2026, 7, 1, 0, 0, 0, 0, time.UTC),
			ExpiresAt: time.Date(2099, 12, 31, 0, 0, 0, 0, time.UTC),
		}}
		materialized := 0.2
		newReq.EffectiveImpact = &materialized

		ev := ChangeEventFromPrevious(cePrevFailing(), &newReq, &prevReq, ceInputs())
		require.NotNil(t, ev)
		assert.Contains(t, ev.ChangeReasons, hdf.EventChangeReasonOverrideAdded)
		count := 0
		for _, r := range ev.ChangeReasons {
			if r == hdf.EventChangeReasonImpactChanged {
				count++
			}
		}
		assert.Equal(t, 1, count, "effectiveImpactChanged must map to a single impactChanged")
		assertValidEvent(t, ev)
	})

	t.Run("base impact change maps to impactChanged", func(t *testing.T) {
		prevReq := ceValid(ecFailingReq())
		newReq := ceValid(ecFailingReq())
		newReq.Impact = 0.3

		ev := ChangeEventFromPrevious(cePrevFailing(), &newReq, &prevReq, ceInputs())
		require.NotNil(t, ev)
		assert.Contains(t, ev.ChangeReasons, hdf.EventChangeReasonImpactChanged)
		assertValidEvent(t, ev)
	})
}

func TestChangeEventFromPrevious_OverrideExpiredAcrossWindow(t *testing.T) {
	// The marquee triage case: a waiver lapsed between observations. Results
	// are identical scans; only the expiry crossing the observation window
	// explains the regression. Deliberately-past expiresAt asserts expiry.
	waived := func() hdf.EvaluatedRequirement {
		req := ceValid(ecFailingReq())
		req.StatusOverrides = []hdf.StatusOverride{stMakeOverride(struct {
			Type      string
			Status    hdf.ResultStatus
			Reason    string
			AppliedAt time.Time
			ExpiresAt time.Time
		}{Type: "waiver", Status: hdf.Passed, ExpiresAt: time.Date(2026, 7, 21, 0, 0, 0, 0, time.UTC)})}
		return req
	}
	prevReq := waived()
	newReq := waived()

	prev := &KeyState{
		EffectiveStatus: "passed",
		EffectiveImpact: 0.5,
		Checksum:        hdf.Checksum{Algorithm: hdf.Sha256, Value: "d11c074ab9131807816013a71d986f1ceb2e5871a8a01dee4043391b7a6bf37b"},
	}
	in := ceInputs()
	in.PrevReferenceTimestamp = "2026-07-20T00:00:00Z"
	in.ReferenceTimestamp = "2026-07-22T14:03:11Z"

	ev := ChangeEventFromPrevious(prev, &newReq, &prevReq, in)
	require.NotNil(t, ev)
	assert.Equal(t, hdf.EventRequirementStateRegressed, ev.State)
	assert.Contains(t, ev.ChangeReasons, hdf.EventChangeReasonOverrideExpired)
	assert.NotContains(t, ev.ChangeReasons, hdf.EventChangeReasonResultChanged,
		"identical scans; the lapsed waiver is the cause")
	assertValidEvent(t, ev)
}

func TestChangeEventFromPrevious_OverrideRemoved(t *testing.T) {
	prevReq := ceValid(ecFailingReq())
	prevReq.StatusOverrides = []hdf.StatusOverride{stMakeOverride(struct {
		Type      string
		Status    hdf.ResultStatus
		Reason    string
		AppliedAt time.Time
		ExpiresAt time.Time
	}{Type: "waiver", Status: hdf.Passed})}
	newReq := ceValid(ecFailingReq())

	prev := &KeyState{
		EffectiveStatus: "passed",
		EffectiveImpact: 0.5,
		Checksum:        hdf.Checksum{Algorithm: hdf.Sha256, Value: "d11c074ab9131807816013a71d986f1ceb2e5871a8a01dee4043391b7a6bf37b"},
	}
	ev := ChangeEventFromPrevious(prev, &newReq, &prevReq, ceInputs())
	require.NotNil(t, ev)
	assert.Equal(t, hdf.EventRequirementStateRegressed, ev.State)
	assert.Contains(t, ev.ChangeReasons, hdf.EventChangeReasonOverrideRemoved)
	assertValidEvent(t, ev)
}
